presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java

```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
+        {
+            long memorySize = 0;
+            for (InternalSingleAccumulator accumulator : accumulators) {
+                accumulator.addInput(inputPage);
+                memorySize += accumulator.getEstimatedSize();
+            }
+            userMemoryContext.setBytes(memorySize);
+            return Optional.empty();
+        }
+
+        @Override
+        public Page getFinalOutput()
+        {
+            List<Block> finalResults = new ArrayList<>();
+            for (InternalSingleAccumulator accumulator : accumulators) {
+                BlockBuilder blockBuilder = accumulator.getFinalType().createBlockBuilder(null, 1);
+                accumulator.evaluateFinal(blockBuilder);
+                finalResults.add(blockBuilder.build());
+            }
+            return new Page(1, finalResults.stream().toArray(Block[]::new));
+        }
+    }
 
-        // output page will only be constructed once,
-        // so a new PageBuilder is constructed (instead of using PageBuilder.reset)
-        PageBuilder pageBuilder = new PageBuilder(1, types);
+    private static class PartialProcessor
+            implements Processor
+    {
+        private final LocalMemoryContext systemMemoryContext;
+        private final List<InternalPartialAccumulator> accumulators;
+        private final IntermediateLayout intermediateLayout;
 
-        pageBuilder.declarePosition();
-        for (int i = 0; i < aggregates.size(); i++) {
-            Aggregator aggregator = aggregates.get(i);
-            BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(i);
-            aggregator.evaluate(blockBuilder);
+        public PartialProcessor(
+                LocalMemoryContext systemMemoryContext,
+                List<InternalAccumulatorFactory> accumulatorFactories, IntermediateLayout intermediateLayout)
+        {
+            this.systemMemoryContext = systemMemoryContext;
+            this.accumulators = accumulatorFactories.stream()
+                    .map(InternalAccumulatorFactory::createPartialAccumulator)
+                    .collect(toImmutableList());
+            this.intermediateLayout = intermediateLayout;
         }
 
-        state = State.FINISHED;
+        @Override
+        public Optional<Page> addInput(Page inputPage)
+        {
+            List<Optional<Block>> partialDistinctMasks = new ArrayList<>(accumulators.size());
+            long memorySize = 0;
+            for (InternalPartialAccumulator accumulator : accumulators) {
+                partialDistinctMasks.add(accumulator.addInput(inputPage));
+                memorySize += accumulator.getEstimatedSize();
+            }
+            systemMemoryContext.setBytes(memorySize);
 
-        Page page = pageBuilder.build();
+            if (!intermediateLayout.hasDistinct()) {
+                return Optional.empty();
+            }
+            return intermediateLayout.buildPartialDistinctRows(inputPage, partialDistinctMasks);
```

Comment:
It looks like when there are lots of duplicate values, we'll be generating very small pages. E.g. it appears that we generate one output page for every input page. Is this the intended behavior?